src/error.rs

```diff
@@ -18,7 +18,73 @@ impl fmt::Display for DatabaseConstraint {
 }
 
 #[derive(Debug, Error)]
-pub enum Error {
+/// The error types for database I/O, connection and query parameter
+/// construction.
+pub struct Error {
+    kind: ErrorKind,
+    original_code: Option<String>,
+    original_message: Option<String>,
+}
+
+pub(crate) struct ErrorBuilder {
+    kind: ErrorKind,
+    original_code: Option<String>,
+    original_message: Option<String>,
+}
+
+impl ErrorBuilder {
+    pub(crate) fn set_original_code(&mut self, code: impl Into<String>) -> &mut Self {
+        self.original_code = Some(code.into());
+        self
+    }
+
+    pub(crate) fn set_original_message(&mut self, message: impl Into<String>) -> &mut Self {
+        self.original_message = Some(message.into());
+        self
+    }
```

Comment:
hehe, the eternal dilemma of `Self` vs `&mut Self` builders

Reply:
yeah... or no chaining at all?